### PR TITLE
6836: elasticsearch filters / public search security

### PR DESCRIPTION
**Pull request by kkoskelin**
ustaxcourt/ef-cms#7401 Revising public search methods to filter search results by utilizing elasticsearch query predicates rather than running the results through an additional filter within the Interactor functions.

Fixing some public entity validation rules that were causing headaches also.

Now public document search interactors also push their search results through a PublicDocketEntry constructor (thanks for the reminder @sutt0n ) as well.

Thanks to @ericsorenson and @codyseibert for their assistance.

### CODE REVIEW

File: shared/src/business/useCases/public/opinionPublicSearchInteractor.js [hunk 1]
```diff
@@ -30,12 +30,14 @@ exports.opinionPublicSearchInteractor = async ({
 
   const rawSearch = opinionSearch.validate().toRawObject();
 
+  // TODO: should these be pushed through a PublicDocument constructor for security?
```
Comment (sutt0n) [inline shared/src/business/useCases/public/opinionPublicSearchInteractor.js]: TODO here -- they should be pushed through `PublicDocketEntry`, i believe, as far as returning an array of whitelisted results

**Review by rachelschneiderman — APPROVED** [review]

File: shared/src/business/useCases/public/opinionPublicSearchInteractor.js (hunk 1 above)
Comment (kkoskelin) [reply shared/src/business/useCases/public/opinionPublicSearchInteractor.js]: DONE

File: shared/src/business/entities/cases/PublicDocketEntry.js [hunk 2]
```diff
@@ -33,50 +31,51 @@ PublicDocketEntry.prototype.init = function init(rawDocketEntry) {
...
+    .description('Indicates the item has been removed from the docket record.'),
+  numberOfPages: DOCKET_ENTRY_VALIDATION_RULE_KEYS.numberOfPages,
+  processingStatus: JoiValidationConstants.STRING.optional(),
   receivedAt: JoiValidationConstants.ISO_DATE.optional(),
   servedAt: JoiValidationConstants.ISO_DATE.optional(),
   servedParties: joi
     .array()
-    .items({ name: JoiValidationConstants.STRING.max(500).required() })
+    .items({
+      name: JoiValidationConstants.STRING.optional().description(
```
Comment (rachaelparris) [inline shared/src/business/entities/cases/PublicDocketEntry.js:78]: Why was this changed from required to optional?

File: shared/src/business/useCases/public/opinionPublicSearchInteractor.js [hunk 3]
```diff
@@ -30,12 +31,17 @@ exports.opinionPublicSearchInteractor = async ({
...
     .advancedDocumentSearch({
       applicationContext,
+      ...rawSearch,
       documentEventCodes: OPINION_EVENT_CODES,
       judgeType: 'judge',
-      ...rawSearch,
+      omitSealed: true,
     });
+
+  return PublicDocketEntry.validateRawCollection(opinions, {
```
Comment (rachaelparris) [inline shared/src/business/useCases/public/opinionPublicSearchInteractor.js]: I'm not sure that this will work here. It looks like we depend on getting back several things from the case combined with the docket entry for the search results, and this will strip out all the case data.

File: shared/src/persistence/elasticsearch/advancedDocumentSearch.js [hunk 4]
```diff
@@ -57,6 +59,11 @@ exports.advancedDocumentSearch = async ({
     });
   }
 
+  if (omitSealed) {
```
Comment (rachaelparris) [inline shared/src/persistence/elasticsearch/advancedDocumentSearch.js:62]: Is there ever a case where we _do_ want sealed cases to be returned?

File: shared/src/persistence/elasticsearch/casePublicSearch.js [hunk 5]
```diff
@@ -48,14 +49,26 @@ exports.casePublicSearch = async ({
 
   let results;
 
+  let excludeSealedDate = {};
+  if (omitSealed) {
```
Comment (rachaelparris) [inline shared/src/persistence/elasticsearch/casePublicSearch.js]: Same as above, is there a case where we do want sealed cases to be returned?

File: shared/src/persistence/elasticsearch/advancedDocumentSearch.js (hunk 4 above)
Comment (kkoskelin) [reply shared/src/persistence/elasticsearch/advancedDocumentSearch.js:62]: Yes, this method gets used internally.

**Review by rachaelparris — APPROVED** [review]
